### PR TITLE
sendgrid domain authentication support

### PR DESCRIPTION
**Pull request by davidcheung**
commitdev/zero#263

### CODE REVIEW

File: modules/sendgrid/variables.tf [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  description = "api_key for sendgrid api, used to authenticate domain once route53 records are created"
+}
+
+variable "zone_name" {
+  description = "route53 zone for CNAME records to be created"
+  type = string
+}
+
+variable "cname" {
+  description = "list of CNAME records to be created for domain authentication"
```
Comment (bmonkman) [inline modules/sendgrid/variables.tf:11]: Maybe indicate here what the two values are in the tuple? (cname, target)

File: modules/sendgrid/variables.tf [hunk 2]
```diff
@@ -0,0 +1,19 @@
+variable "sendgrid_api_key_secret_name" {
+  description = "api_key for sendgrid api, used to authenticate domain once route53 records are created"
+}
+
+variable "zone_name" {
+  description = "route53 zone for CNAME records to be created"
+  type = string
+}
+
+variable "cname" {
```
Comment (bmonkman) [inline modules/sendgrid/variables.tf:10]: should be `cnames`

File: modules/sendgrid/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# Sendgrid
+
+Authenticates domain with sendgrid, assumes CNAME required are provided in variables
```
Comment (bmonkman) [inline modules/sendgrid/README.md:3]: `assumes CNAME required are provided in variables` is redundant based on the rest of the docs.

File: modules/sendgrid/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+## Notes
+
+<!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
+## Requirements
+Assumes domain is already awaiting authentication, this created the route53 entries
```
Comment (bmonkman) [inline modules/sendgrid/README.md:9]: created -> creates

File: modules/sendgrid/sendgrid-external-datasource.sh [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+## Posts to create if not already exist
+## This condition is a simple way of making sure this script is idempotent, should yield same result no matter how many times we call it
+if [ "${MATCHING_DOMAIN_COUNT}" -lt "1" ]; then
+  curl -s -XPOST "${SENDGRID_ENDPOINT}whitelabel/domains" -H "Authorization:Bearer ${SENDGRID_API_KEY}" --data "{\"domain\": \"${DOMAIN}\"}" > /dev/null
+fi
+
+## Print our CNAMES for our target domain
+## IMPORTANT: must print out map[string]string in JSON format, where all values must be string(CANNOT be int/numbers/boolean)
+JQ_CNAMES_QUERY=".[] | select( .domain == \"${DOMAIN}\" ) | { domain_id : .id|tostring } +  (.dns | with_entries(.key |=\"\(.)_host\" | .value |=.host ) )+  (.dns | with_entries(.key |=\"\(.)_value\" | .value |=.data ) )"
```
Comment (bmonkman) [inline modules/sendgrid/sendgrid-external-datasource.sh:18]: Epic jq!

File: modules/sendgrid/variables.tf [hunk 6]
```diff
@@ -0,0 +1,14 @@
+variable "zone_name" {
+  description = "route53 zone for CNAME records to be created"
+  type = string
+}
+
+variable "sendgrid_api_key_secret_name" {
+  description = "api_key for sendgrid api, used to authenticate domain once route53 records are created"
```
Comment (bmonkman) [inline modules/sendgrid/variables.tf:7]: How about something like "name of the secret in AWS secret manager that contains the sendgrid API key"? Also could you please run the pre-commit hook so this stuff gets populated into the readme?

File: modules/sendgrid/README.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+## Notes
+This implementation relies on a bash-script to fetch the CNAMEs required for sendgrid domain authentication, the bashscript must be implemented in an idempotent way where it always returns the same result for terraform state.
+
+<!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
+## Requirements
+- jq
+- curl	No requirements.
+- Api key to be stored in AWS secret manager
+- HostedZone to be properly configured as root of domain in Route53
```
Comment (davidcheung) [inline modules/sendgrid/README.md:13]: should these be moved to `## Notes` instead of requirements since theyre generated

Comment (bmonkman) [reply modules/sendgrid/README.md:13]: Yep!

Comment (bmonkman) [reply modules/sendgrid/README.md:13]: Otherwise next time someone runs the pre-commit hook that will be wiped out.

**Review by bmonkman — APPROVED** [review]

**Pull request merged**